peps/pep-0734.rst

```diff
@@ -0,0 +1,800 @@
+      (e.g. with imports) before later performing some repeated task.
+
+Comparison with builtins.exec()
+-------------------------------
+
+``Interpreter.exec()`` is essentially the same as the builtin
+``exec()``, except it targets a different interpreter, using that
+interpreter's distinct runtime state.
+
+Here are the relevant characteristics of the builtin ``exec()``,
+for comparison:
+
+* It runs in the current OS thread and pauses whatever was running there,
+  which resumes when ``exec()`` finishes.  No other threads are affected.
+  (To avoid pausing the current thread, run ``exec()``
+  in a ``threading.Thread``.)
+* It executes against a namespace, by default the ``__dict__`` of the
+  current module (i.e. ``globals()``).  It uses that namespace as-is
+  and does not clear it before or after.
+* When code is run from the command-line (e.g. ``-m``) or the REPL,
+  the "current" module is always the ``__main__`` module
+  of the target (main) interpreter.
+* It discards any object returned from the executed code::
+
+   def func():
+       global spam
+       spam = True
+       return 'a returned value'
+
+   ns = {}
+   res = exec(func.__code__, ns)
+   # Nothing is returned.  The returned string was discarded.
+   assert res is None, res
+   assert ns['spam'] is True, ns
+
+* It propagates any uncaught exception from the code it ran.
+  The exception is raised from the ``exec()`` call in the thread
+  that originally called ``exec()``.
+
+The only difference is that, rather than propagate the uncaught
+exception directly, ``Interpreter.exec()`` raises an
+``interpreters.RunFailedError`` with a snapshot
+(``traceback.TracebackException``) of the uncaught exception
+(including its traceback) as the ``__cause__``.  That means if the
+``RunFailedError`` isn't caught then the full traceback of the
+propagated exception, including details about syntax errors, etc.,
+will be displayed.  Having the full traceback is particularly useful
+when debugging.
+
+Communicating Between Interpreters
+----------------------------------
+
+The module introduces a basic communication mechanism through special
+queues.
+
+There are ``interpreters.Queue`` objects, but they only proxy
+the actual data structure: an unbounded FIFO queue that exists
+outside any one interpreter.  These queues have special accommodations
+for safely passing object data between interpreters, without violating
+interpreter isolation.  This includes thread-safety.
+
+As with other queues in Python, for each "put" the object is added to
+the back and each "get" pops the next one off the front.  Every added
+object will be popped off in the order it was pushed on.
+
+Only objects that are specifically supported for passing
+between interpreters may be sent through a ``Queue``.  Note that the
+actual objects aren't sent, but rather their underlying data.
+However, the popped object will still be identical to the original.
```

Comment:
"Identical" is probably not the term you're looking for here -- that would imply the two have the same address. I'd use equivalent or equal instead.